### ISSUE TITLE
icp_blake3_impl tunable broken

### ISSUE THREAD

**Issue by freqlabs** [openzfs/zfs#13647]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| any
Distribution Version	| any
Kernel Version	| any
Architecture	| any
OpenZFS Version	| master @ 1d3ba0bf01020f5459b1c28db3979129088924c0

### Describe the problem you're observing

The icp_blake3_impl module parameter is updated incorrectly at run time, so only "cycle" and "fastest" will reflect being selected.

Having looked at the code (I am in the process of exposing this tunable for FreeBSD), it seems the parameter will only be displayed correctly if it is set at module load time rather than at run time. There is some overly-complicated looking IMPL_PARAM/IMPL_USER distinction that is to blame. It should probably work a lot more like in zfs_fletcher_4_impl in module/zcommon/zfs_fletcher.c.

### Describe how to reproduce the problem

```
# cat /sys/module/zfs/parameters/icp_blake3_impl
cycle [fastest] generic sse2 sse41 %
# echo sse41 > /sys/module/zfs/parameters/icp_blake3_impl
# cat /sys/module/zfs/parameters/icp_blake3_impl
cycle fastest generic sse2 sse41 %
# echo fastest > /sys/module/zfs/parameters/icp_blake3_impl
# cat /sys/module/zfs/parameters/icp_blake3_impl
cycle [fastest] generic sse2 sse41 %
```

**Comment by jumbi77** [comment]
ping @mcmilk (in case you are interested).

**Comment by rincebrain** [comment]
You know, I swear this worked when I tested it. (Of course, I believe that was before IMPL_PARAM/IMPL_USER was added...I don't recall seeing any discussion about it in the PR. Hopefully @mcmilk can clarify the problem this is solving, rather than someone just ripping it out and having to find out later...

**Comment by mcmilk** [comment]
I am on holiday currently. But I will fix this, when I am home again - in about one week.

**Comment by freqlabs** [comment]
In that case, it may be useful to incorporate or at least take into consideration the changes I've made here to support the tunable on FreeBSD: https://github.com/truenas/zfs/tree/blake3-tunable

In general FreeBSD doesn't use any icp code, except for edonr and now blake3, so the icp tunables not being implemented for FreeBSD hasn't been a problem until now.

By the way, this all came to my attention while investigating a report of BLAKE3 causing pool corruption:
https://lists.freebsd.org/archives/freebsd-current/2022-July/002281.html

**Comment by mcmilk** [comment]
I am working on [SHA2 SIMD code](https://github.com/mcmilk/sha2-testing)... I have to read more about the different ways of using the checksumming code in OpenZFS for Linux vs FreeBSD.
Maybe we can work together on this ... I will start a PR in the end of the next week I think.

**Comment by mcmilk** [comment]
The generic implementation interface is a bit more complex to do... so I will just fix this issue via extra PR... and then implement the SHA2 reworking.

Then, when the SHA2 stuff is in, we have these three algos within the ICP codebase which also get used bei FreeBSD:
- BLAKE3
- SHA2
- EdonR

Is the  `icp/illumos-crypto.c` stuff of interest for FreeBSD OpenZFS?
In the codebase of illumos is ZFS very highly integrated, so I have no idea why we have some file called `illumos-crypto.c` with an `ifdef __FreeBSD__` :-)

I will also reuse this here: https://github.com/truenas/zfs/tree/blake3-tunable - thanks for this beginning.

**Comment by freqlabs** [comment]
The `ifdef __FreeBSD__` in `icp/illumos-crypto.c` is a mistake left by ad9e767657c3d3c0ec9e09a17f6732bcbf915401 when ICP was welded into the zfs module for Linux. They apparently didn't realize FreeBSD does not use that file in the kernel.

**Comment by mcmilk** [comment]
@freqlabs - can this issue be closed?

**Comment by behlendorf** [comment]
Yes, I believe it can.  Resolved by #13725.
